rust/wal3/README.md

```diff
@@ -539,44 +580,40 @@ object store.  This becomes expensive as the number of logs grows, especially if
-
-One potential straw-man approach considered would be to hash each log's prefix and then insert into
-a prefix of dirty logs.  It looks something like this:
+Conceptually, we're essentially looking for a way to aggregate the information about which
+logs were written when, so that we can compact and garbage collect logs and keep system resource
+usage for cleaning up proportional to the cost of making the mess.
 
-```text
-wal3/dirty/ea8cd5d4f1b2ed79d8415c832e2fd2cef87e6ebcecde12aa01b2ccc5aceaab55
-wal3/dirty/ea47843f7d677b67d7364c2b34322770988e91e2b21df4d7c98064dfb14ae68d
-```
+The insight we'll make use of here is, essentially, that the logs written by a single server can be
+summarized with a fraction of the resources of the server.  For starters, we don't need to write as
+much data to say a log has data as we write to that log.  If each append to a log is approximately
+4 kB+ (a reasonable document size with a vector), we can track the dirty log by name using at most
+48 B; an 85x reduction.  But it goes further---we only need to persist that 48B record once per
+fragment write, not once per append.
 
-The problem with this approach is that listing `wal3/dirty` becomes expensive as things grow.  With
-1e9 collections, simply knowing which are dirty would cost $10k to LIST.  What if we knew every
-potential log name and could enumerate them?  We could do a GET/HEAD on each dirty "bit" to see if
-it exists in object storage.  This comes out to $5k per pass for 1e9 objects, but violates the
-requirement that work be proportional to the data written.
+This means that we can track the dirty logs with a single 48B record per fragment.  Where to put
+that record?  We need some way to record them as they happen and then scan/roll-up the records into
+a summary of dirty logs at all times.
 
-The observation to have is that if collection dirty bits are 1:1 with files in object storage, then
-there will be O(dirty collections) files to list.  One straw man approach considered was to make a
-hierarchy in S3 like a B+-tree:
+Viewed another way, we have a stream of events that we need to record approximately in order.
 
-```text
-wal3/dirty/ea/8c/d5/d4/f1b2ed79d8415c832e2fd2cef87e6ebcecde12aa01b2ccc5aceaab55
-wal3/dirty/ea/47/84/3f/7d677b67d7364c2b34322770988e91e2b21df4d7c98064dfb14ae68d
-```
+What if we just put it in another log?  wal3 is already configured to dynamically batch and write
+log data to object storage in an efficient manner.  Further, we know a single machine's multiplicity
+of logs can be sustained by the throughput of a single log by the math above.
 
-This strictly increases the number of list operations, but brings a useful property:  We can
-partition the space of dirty bits so that multiple scanners can pass over them in parallel.  This
-doesn't reduce the cost of finding dirty logs, but does make it faster.
+The protocol for discovering dirty logs, then, is to write to a "dirty" log and roll-up the dirty
```

Comment:
probably useful to also talk about failure modes here. For e.g. what if the node crashes while the dirty log is still being batched in memory